Merge Tensor and Variable.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28620 Merge Tensor and Variable.**
* #29047 Print which output didn't have dependence.

All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

ezyang committed:
Merge Tensor and Variable.

All Tensors are Variables now, they just happen to have requires_grad=False.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
Status update: I'm still munching through the bugs. One thing I've noticed is that tracing of constants is broken after this change

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
Tentatively bug free and ready for review

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

albanD commented on c10/core/TensorImpl.cpp:
```diff
@@ -60,7 +52,7 @@ TensorImpl::TensorImpl(Storage&& storage, TensorTypeSet type_set, const caffe2::
       data_type_(data_type),
       device_opt_(device_opt),
-      type_set_(type_set.remove(TensorTypeId::VariableTensorId)) {
+      type_set_(type_set.add(TensorTypeId::VariableTensorId)) {
```
This happens in a few places.

ezyang replied:
The invariant is specifically, every Tensor has a type set which contains VariableTensorId. The invariant is not, every occurrence of TensorTypeSet contains a VariableTensorId (it's quite difficult to arrange for this to be the case, as we have a thread local TensorTypeSet and thread locals must be initialized to all zeros for fast TLS to be used.)

albanD replied:
Is this invariant described explicitly somewhere?

ezyang replied:
Yes, in the PR description, and I've also added it to `type_set_` documentation

gchanan replied:
I think you should comment here too.

albanD commented on caffe2/python/workspace_test.py:
```diff
@@ -299,10 +299,8 @@ def testTorchInterop(self):
-        # the underlying data tensor is not properly reshaped :(
         np.testing.assert_array_equal(
-            workspace.FetchBlob("foo"), np.array([2,2,777,2]))
+            workspace.FetchBlob("foo"), np.array([2,2,777,2,777]))
```
Does that mean BC? Or just bug fixing?

ezyang replied:
This is technically BC-breaking, but I am not too concerned about it:
1. The test explicitly points out that they wish the resize was reflected in the blob (it just wasn't because of the shallow copy)
2. The BC-breakage is specifically related to PyTorch-Caffe2 interop, which doesn't have too many external users

You could call it BC, you could call it bug fixing.

albanD commented on torch/csrc/autograd/variable.h:
```diff
@@ -518,9 +518,6 @@ inline Variable make_variable(
     at::Tensor data,
```
What is this function supposed to do now? The comment seems out of data as well.

ezyang replied:
I'm planning in a follow up to delete all occurrences of `make_variable`. But I think in the current codebase it kind of plays the same role as the legacy `Variable` constructor in Python.

albanD commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -667,8 +667,12 @@ at::Tensor PythonArgs::tensor_slow(int i) {
+    tensor = scalar_to_tensor(scalar);
+    tensor.unsafeGetTensorImpl()->set_wrapped_number(true);
+  }
   return autograd::make_variable(tensor);
```
The `autograd::make_variable()` is kept here but removed from the change to `internal_new_from_data()` in `tensor_new.cpp`.
Is there a reason? Should we keep using `make_variable()` going forward?

ezyang replied:
I think every call that looks like `make_variable(x)` can be soundly replaced with `x`, but I haven't carefully verified this. In `internal_new_from_data` it looks like I accidentally deleted the `make_variable` (but it looks like it didn't make a difference)

albanD replied:
I didn't look into `make_variable` in great details but it seems to be doing non-trivial stuff. No?

ezyang replied:
Most of what it is doing is making a shallow copy of the tensor and setting some properties (like `requires_grad`) appropriately on the shallow tensor. Some of the complication is that when you only have one reference to the tensor, you just do that modification in-place.

In the new world, I think most occurrences of `make_variable(x)` with no arguments are a no-op, because `x` is supposed to not require grad already. Occurrences of `make_variable(x, /*requires_grad*/ true)` probably could have been replaced by just by initializing the variable correctly in the first place. It's hard to say and will need to be done on a case-by-case basis.

albanD replied:
Sounds good.

ezyang commented on aten/src/ATen/InitialTensorOptions.h:
```diff
@@ -9,7 +9,7 @@ namespace at {
 inline TensorOptions initialTensorOptions() {
   return TensorOptions(kCPU).dtype(kFloat).layout(kStrided)
-                            .requires_grad(false).is_variable(false);
+                            .requires_grad(false);
```
I don't think `requires_grad(false)` is actually needed here. But I didn't change it, as the only modification I was making here was just to delete `is_variable`.

TensorOptions `requires_grad` is the C++ analogue of `requires_grad=True` keyword argument in `torch.empty`, etc. It's just a possible kwarg that will affect whether or not the newly returned tensor from the function in question requires grad or not. Requires grad on AutogradMeta is the thing that actually tells you if a variable requires grad or not.

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -252,7 +242,6 @@ void TensorImpl::set_autograd_meta(std::unique_ptr<c10::AutogradMetaInterface> a
   // NB: autograd_meta may be null!  That just means it's the default
   // constructor
   autograd_meta_ = std::move(autograd_meta);
-  type_set_ = type_set_.add(TensorTypeId::VariableTensorId);
```
Yes, per the invariant that the type set of every Tensor already contains VariableTensorId

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

albanD commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -104,7 +104,7 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
       at::Tensor tensor = ([&]() {
         at::AutoNonVariableTypeMode non_var_type_mode(true);
-        return at::empty(sizes_, at::TensorOptions(options).device(at::kCPU).is_variable(false));
+        return at::empty(sizes_, at::TensorOptions(options).device(at::kCPU));
```
Should these be guarded with `AutoNonVariableTypeMode` ?
Is that the right guard to use to perform non-autograd stuff in cpp? Or `NoGradGuard` is the right thing to use?

ezyang replied:
The situation with `AutoNonVariableTypeMode` and `NoGradGuard` is very delicate right now. In the terminal state, these two guards would not be distinguishable. However, because tracing, profiling, and version counter logic lives inside VariableType implementations, these guards have different functions: NoGradGuard has the same function as `no_grad`, but we will still execute VariableType implementations; `AutoNonVariableTypeMode` will bypass VariableType entirely.

Most of the insertions of AutoNonVariableTypeMode in this patch are to solve problems with tracing, so `NoGradGuard` wouldn't actually solve our problem (because you are still going to hit the tracing code.) The most common situation looks something like this:

```
Tensor t = at::empty(...);
// do some raw data access on t, e.g., set storage or mutate data pointer
Variable v = make_variable(t);
```

Without the AutoNonVariableTypeMode guard this will trace incorrectly because `at::empty` now generates an `at::empty` in the trace, but then we fail to trace the raw data access. Previously it would have worked as `at::` would have returned a non-Variable which doesn't trigger tracing.

albanD replied:
From a new user point of view that write brand new cpp code.
- The `NoGradGuard` should be used just like `no_grad()` in python.
- `AutoNonVariableTypeMode` is an internal thing that should NOT be used. Do like python: use requires_grad(false) + detach() to get what you want.

For existing code: people should not be doing factory functions so this is not BC.

ezyang replied:
Yes, that sounds accurate to me

ezyang replied:
I'm going to beef up the documentation on this guard

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

albanD commented on aten/src/ATen/cpp_custom_type_hack.h:
```diff
@@ -26,6 +26,9 @@ T& cast(const Tensor& packed) {
 
 template <typename T>
 Tensor create(std::unique_ptr<T> ptr, TensorOptions options) {
+  // None of this should trace, so turn off Variable handlign
```
nit: handling

albanD commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -20,6 +20,8 @@
 #include <ATen/NamedTensorUtils.h>
 #include <ATen/core/EnableNamedTensor.h>
 
+#include <ATen/core/grad_mode.h>
```
Is this include necessary?

ezyang replied:
Hmm, I guess not!

albanD commented on c10/core/TensorImpl.cpp:
```diff
@@ -29,27 +29,19 @@ const char * const TensorImpl::err_msg_tensor_metadata_change_not_allowed =
+  // is not so easy to fix right now because the mutable counterpart of
+  // this function must keep working so that "x.grad() = ..." keeps working
+  // (part of public API).
+  if (!autograd_meta_) impl::GetAutogradMetaFactory()->undefined_tensor();
```
I think I'm missing something here. I can't find an `undefined_tensor()` method on the factory.

ezyang replied:
Factory, not interface.

```
struct C10_API AutogradMetaFactory {
  virtual ~AutogradMetaFactory() = default;
  virtual std::unique_ptr<AutogradMetaInterface> make() const = 0;
  // This method is the dumbest method.  But I don't have access
  // to Tensor (not TensorImpl) which is undefined in this header.
  virtual const at::Tensor& undefined_tensor() const = 0;
};
```

albanD replied:
Why do you call this function and do nothing with the result?

ezyang replied:
oooh, I'm supposed to return it. good catch

albanD commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -277,7 +276,11 @@ Tensor internal_new_from_data(
+  // This exists to prevent us from tracing the call to empty().  The actual
+  // autograd code doesn't really matter, because requires_grad is always false
+  // here.
+  at::AutoNonVariableTypeMode guard;
```
nit: `guard_` for consistency

ezyang replied:
I probably should change the other guards around. Trailing underscore reserved for properties on C++ classes...

albanD reviewed (approved):
LGTM, only some details left.
It might also be interesting to include a documentation note on the difference between the guards as discussed in the PR. But that can be part of a later PR.

yf225 commented on aten/src/ATen/core/LegacyTypeDispatch.h:
```diff
@@ -62,4 +62,10 @@ struct CAFFE2_API AutoNonVariableTypeMode {
 };
 
+namespace impl {
+inline bool variable_is_excluded() {
```
I feel that the function name is a bit unintuitive (it wasn't immediately clear to me at first sight what it is checking). Perhaps we could name it something like `is_non_variable_type` ? (not sure if it is better)

ezyang replied:
The trouble is, "what", precisely, is the non variable type? Not sure there is much of an improvement here either, alas.

gchanan replied:
Does this only effect dispatch now?  If that's true, "variable_excluded_from_dispatch" is probably clearer.

ezyang replied:
Yes, it only excludes dispatch. I suppose I like the new name.

yf225 commented on c10/core/TensorOptions.h:
```diff
@@ -228,14 +224,6 @@ struct C10_API TensorOptions {
-    TensorOptions r = *this;
-    r.set_is_variable(is_variable);
-    return r;
-  }
```
We probably need to declare BC-breaking on the removal of the `is_variable` accessors because they are public API (which should be a minor issue because I don't think libtorch end users are actually using it).

yf225 commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -277,7 +276,11 @@ Tensor internal_new_from_data(
+  // autograd code doesn't really matter, because requires_grad is always false
+  // here.
+  at::AutoNonVariableTypeMode guard;
+  auto tensor = at::empty(sizes, at::initialTensorOptions().dtype(inferred_scalar_type).pinned_memory(pin_memory));
```
should we write
```cpp
at::Tensor tensor = ([&]() {
  at::AutoNonVariableTypeMode guard;
  return at::empty(sizes, at::initialTensorOptions().dtype(inferred_scalar_type).pinned_memory(pin_memory));
}
```
or does the `at::AutoNonVariableTypeMode` guard apply to the rest of code as well?

ezyang replied:
It applies to the rest of the code. At the end of the day I want this entire tensor to come out with NO tracing state applied to it, so it gets handled as a constant.

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/core/Tensor.cpp:
```diff
@@ -13,18 +13,15 @@ void Tensor::enforce_invariants() {
-          impl_->storage_initialized(),
-          "Partially-initialized tensor not supported by at::Tensor");
-    }
+    AT_ASSERTM(
```
nit: TORCH_CHECK?

ezyang replied:
I plead preexisting code XD

gchanan replied:
I think AT_ASSERT is actually correct here (ok, you could probably argue about sparse).  But the other use cases are weird enough that it's not really considered part of the public API.

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 1aac7f1:
* 1/1 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=1aac7f1d350e00a00c7372de353cc112dbfc4269).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

ezyang committed:
Update on "Merge Tensor and Variable."


All Tensors are Variables now, they just happen to have requires_grad=False. Tensors ALWAYS have `VariableTensorId` in their type set.

When constructing this patch, I had to make decisions about what I would fix in this patch, and what I would leave for follow up PRs. Here is the cleanup that happens in this patch:

- The `is_variable` property is removed from TensorOptions. I removed this immediately because unlike Tensor::is_variable, TensorOptions::is_variable doesn't respect our VariableTensorId thread-local state. This means that there were a bunch of places where TensorOptions::is_variable was false, which is obviously bogus in the world when tensor and variable are merged. Instead of keeping the method as a function that always returns true, I just opted to remove it entirely (it's not public API.) All places we set `is_variable` are deleted.
  - Knock on effect: there is no longer a separate DeprecatedTypeProperties for the variable and non-variable versions of type.
  - Knock on effect: instead of asserting on TensorOptions::is_variable, instead we just test `at::impl::variable_is_excluded()`
- There is now only one copy of the cuDNN RNN dropout cache, not two (I'm not sure why we had two to begin with)

Some cleanup that doesn't happen in this patch:
- Eliminating unnecessary uses of `make_variable`
- Eliminating `Tensor::is_variable`

The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

- `torch.tensor` must turn off VariableType before invoking `at::empty` to construct the tensor, as it subsequently does direct data access
- `tensor_slow` (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling `scalar_to_tensor` so the scalar gets traced as constant, rather than as a call to `scalar_to_tensor`.

Honestly, these are all giant hacks, and should be replaced with a more specialized guard that just toggles tracing.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171156](https://our.internmc.facebook.com/intern/diff/D18171156)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/Context.h:
```diff
@@ -142,22 +142,22 @@ CAFFE2_API Allocator* getCPUAllocator();
 static inline DeprecatedTypeProperties& getNonVariableDeprecatedTypeProperties(Backend p, ScalarType s) {
   return globalDeprecatedTypePropertiesRegistry().getDeprecatedTypeProperties(
-      p, s, /*is_variable*/false);
+      p, s);
```
umm...this seems super misleading to leave this function as is.

ezyang replied:
Trying to keep this PR on the smaller side :)

gchanan replied:
you could put up a pre-req PR.

ezyang replied:
#29203

gchanan commented on aten/src/ATen/core/DeprecatedTypeProperties.h:
```diff
@@ -19,8 +19,8 @@ class Tensor;
  public:
-  DeprecatedTypeProperties(Backend backend, ScalarType scalar_type, bool is_variable)
-    : backend_(backend), scalar_type_(scalar_type), is_variable_(is_variable) {}
+  DeprecatedTypeProperties(Backend backend, ScalarType scalar_type)
```
wait, is it true that we haven't actually deprecated "DeprecatedTypeProperties" (i.e. via the public accessor, `tensor.type()`?

ezyang replied:
I think we forgot to put the deprecated marker on it. See also https://github.com/pytorch/pytorch/pull/25306

gchanan replied:
Thanks!  I filed https://github.com/pytorch/pytorch/issues/29161.

gchanan commented on aten/src/ATen/core/LegacyTypeDispatch.h:
```diff
@@ -50,7 +50,34 @@ CAFFE2_API LegacyTypeDispatch& globalLegacyTypeDispatch();
+//
+// TODO: Since `torch::NoGradGuard` serves the same purpose in libtorch, we should
+// merge these two thread-local guards.  However, NoGradGuard does something
+// subtly different: it turns of gradient recording, but DOES NOT skip
```
of -> off.

gchanan commented on aten/src/ATen/core/LegacyTypeDispatch.h:
```diff
@@ -50,7 +50,34 @@ CAFFE2_API LegacyTypeDispatch& globalLegacyTypeDispatch();
+// when enabled will cause `legacyTensorType()` and `getType()` to always return
+// non-Variable type, even if the tensor being called on is a variable.
+//
+// TODO: Since `torch::NoGradGuard` serves the same purpose in libtorch, we should
```
I'd say "serves _almost_ the same purpose" or something -- this is something we got wrong before and the statement as it is now is definitely misleading.

gchanan commented on aten/src/ATen/cpp_custom_type_hack.h:
```diff
@@ -26,6 +26,9 @@ T& cast(const Tensor& packed) {
 template <typename T>
 Tensor create(std::unique_ptr<T> ptr, TensorOptions options) {
+  // None of this should trace, so turn off Variable handling
+  at::AutoNonVariableTypeMode guard;
```
umm...errr...I guess this is right?  But the tracing seems like a bit of an implementation detail, i.e. what you really want is some torch-level VariableType thing that does the "unpacking / setting the guards".  Which suggests this is at the wrong level?

Put another way -- I'm assuming because you put this here it was necessary, but what's the backtrace that makes this happen?

ezyang replied:
From PR description

> The most subtle part of this patch is retaining tracing behavior: the fact that everything is a Variable means that more code gets routed to VariableType than before; this can change traces. I identified two places where we didn't appropriately turn off VariableType, mostly factory functions:

> * torch.tensor must turn off VariableType before invoking at::empty to construct the tensor, as it subsequently does direct data access
> * tensor_slow (invoked when you pass a Python scalar to a tensor argument) must turn off VariableType before calling scalar_to_tensor so the scalar gets traced as constant, rather than as a call to scalar_to_tensor.

I believe, concretely, for this case, quantized tests that utilized (1) tracing and (2) custom cpp type hack failed without this guard, because without the guard we spew an unnecessary `at::empty` into the trace.

gchanan replied:
I'm a bit confused with the wording -- this is a third case?

ezyang replied:
Yeah, it looks like I forgot to update the PR description with more cases as I found them.

gchanan commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -71,7 +71,7 @@ Tensor values_sparse(const Tensor& self) {
 
 SparseTensor new_sparse(const TensorOptions& options) {
-  AT_ASSERT(!options.is_variable());  // TODO: remove this when Variable and Tensor are merged
+  TORCH_INTERNAL_ASSERT(impl::variable_is_excluded());
```
is it true that these statements are equivalent?  I believe that's true for `tensor.is_variable()` and `impl::variable_is_excluded()` but I'm not sure about `options.is_variable()`.

ezyang replied:
No, they're not equivalent. `TensorOptions::is_variable` doesn't consult the TLS at call site; there is a value stored in TensorOptions. I don't think that value is seeded from TLS either; it was manually twiddled before (all that code is gone now).

I think the new code makes sense without appealing to the old code. The analogy is with the other factory functions, which also assert variable is excluded.

gchanan commented on aten/src/ATen/templates/Functions.h:
```diff
@@ -28,6 +28,7 @@ inline Tensor from_blob(
     IntArrayRef strides,
     const std::function<void(void*)>& deleter,
     const TensorOptions& options = {}) {
+  AutoNonVariableTypeMode guard;
```
this seems "wrong" in the sense that I would expect AutoNonVariableTypeMode only at the torch level.  I.e. I think this is actually correct, but not maintainable, this should be going through variable_factories.

ezyang replied:
`from_blob` is special because there's no torch level binding code for it, and it's manually bound in TensorFunctions. So there isn't anywhere else to put it: user level C++ calls will directly drop here. So yes, it's "unmaintainable", but only in the same sense all of the manually bound functions in PyTorch are unmaintainable. It's something you'll have to remember to do, along with setting device guards, etc. Probably something we should talk about in the docs.

One question is if plans to separate tracing into its own mode help make this more maintainable. I don't think it does though, for precisely the reason that it is manually bound.

gchanan replied:
But why is it manually bound?  We have a bunch of related support at the torch level for things like this (e.g. torch::tensor) -- what's special about this one?

ezyang replied:
It looks like torch::tensor is manually bound too. Actually, the situation in torch::tensor is quite similar, except that the call to AutoNonVariableTypeMode happens inside torch namespace rather than at namespace.

`torch::tensor` is defined in tools/autograd/templates/variable_factories.h and the action happens in `torch/csrc/api/include/torch/detail/TensorDataContainer.h`. There isn't really any reason this class has to live in torch versus at; it is operating at basically the same level of abstraction as `from_blob`

gchanan commented on c10/core/TensorImpl.cpp:
```diff
@@ -29,27 +29,19 @@ const char * const TensorImpl::err_msg_tensor_metadata_change_not_allowed =
-  } else {
-    AT_ERROR("grad is not implemented for Tensor");
-  }
+  if (!autograd_meta_) autograd_meta_ = impl::GetAutogradMetaFactory()->make();
```
why does this not check the excluded TLS?

ezyang replied:
https://github.com/pytorch/pytorch/pull/28620#discussion_r342234685

gchanan commented on c10/core/TensorImpl.cpp:
```diff
@@ -29,27 +29,19 @@ const char * const TensorImpl::err_msg_tensor_metadata_change_not_allowed =
-  } else {
-    AT_ERROR("grad is not implemented for Tensor");
-  }
+  // Yes, I know this looks really weird.  But I don't really have a choice as
```
here too.

ezyang replied:
https://github.com/pytorch/pytorch/pull/28620#discussion_r342234685

gchanan commented on c10/core/TensorImpl.cpp:
```diff
@@ -235,7 +227,6 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
 AutogradMetaInterface::~AutogradMetaInterface() {}
 
 void TensorImpl::set_requires_grad(bool requires_grad) {
-  TORCH_INTERNAL_ASSERT(type_set_.has(TensorTypeId::VariableTensorId), "set_requires_grad is not implemented for Tensor");
```
for these why don't you check that we aren't in the excluded scope?

ezyang replied:
Because it's inconsistent with the old code. The old code only tests `type_set_` and doesn't respect TLS; the new code eliminates the `type_set_` test entirely (because all tensors are variables) and, as before, doesn't respect TLS (which would be the test for excluded scope).

I suppose there is a broader question which is whether or not AutoNonVariableTypeMode should "disable" all autograd operations while it is set. I am not too sure about how to answer this question, but the plan on record is that we make it valid to bypass VariableType entirely for operations which don't require grad. If that becomes the terminal meaning of AutoNonVariableTypeMode (aka `no_grad`), then we absolutely should permit `set_requires_grad` inside a scope like this, in the same way that you can still set requires_grad inside a no_grad block in Python.

gchanan commented on c10/core/TensorOptions.h:
```diff
@@ -313,17 +301,6 @@ struct C10_API TensorOptions {
   }
 
-  /// Returns the `is_variable` property of the `TensorOptions`.
-  bool is_variable() const noexcept {
```
I'm not convinced we should release like this (it's good that you did this to make sure internal things don't call this).  I'm actually not 100% sure what the equivalent should be (how does variable_factories work?), but it might be something like change the default and assert if a user ever tries to set it to False.

ezyang replied:
> I'm actually not 100% sure what the equivalent should be (how does variable_factories work?)

Post this PR, it's always true. You always get a variable no matter what you do.

I agree this is modestly BC-breaking (and I had to fix some sites in fbcode where `is_variable` was being set), but I think this method never really made any sense for end users (it was more of a situation where people sprinkled it in just get things to work, without understanding what it did) and I think it's clearer if we just force people to delete it from their code. At least, there were very *few* fbcode fixes I had to do, and I use that as a proxy for external usage.

ezyang commented:
@gchanan I read your review and I didn't see any major issues raised by you. If it's OK, I'm going to opt to land this diff as is and fix your comments in follow up, to help unblock @dreiss who needs this PR sooner rather than later.
